From 8a8e2590d1c05348ed6f3155cf4038a8c7a6ba48 Mon Sep 17 00:00:00 2001
From: Max Ihlenfeldt <max@igalia.com>
Date: Tue, 16 Jan 2024 12:29:30 +0000
Subject: [PATCH 09/11] Adjust the Rust build to our needs

1. Set `RUSTC_BOOTSTRAP=1` environment variable when calling rustc, to
   avoid the need of a nightly Rust toolchain.
2. Copy the target.json file which rustc needs to compile for OE's
   custom target triple.
3. Add a rust_target_triple_vendor_for_target GN arg. OE builds the
   Rust libraries for the target by using TARGET_VENDOR as the vendor
   part for the Rust "target triple", but Chromium assumes the vendor to
   always be "-unknown". This new arg lets us override the default when
   building target code.
4. Use host rust compiler when with host_system_allocator_toolchain
5. Remove arm32 bit specific clang flags from rust compiler invocation

Upstream-Status: Inappropriate [specific to our build setup]
Co-authored-by: Caner Altinbasak <cal@brightsign.biz>
Signed-off-by: Max Ihlenfeldt <max@igalia.com>
---
 build/config/rust.gni            | 28 ++++++++++++++++++----------
 build/rust/filter_clang_args.py  |  5 +++++
 build/rust/rustc_wrapper.py      |  1 +
 build/rust/std/BUILD.gn          | 28 +++++++++++++++++++++-------
 build/rust/std/find_std_rlibs.py | 16 +++++++++++-----
 5 files changed, 56 insertions(+), 22 deletions(-)

diff --git a/build/config/rust.gni b/build/config/rust.gni
index bd18d13119..e260576fba 100644
--- a/build/config/rust.gni
+++ b/build/config/rust.gni
@@ -95,6 +95,11 @@ declare_args() {
   # a platform. Mostly applicable to Windows, where new versions can handle ANSI
   # escape sequences but it's not reliable in general.
   force_rustc_color_output = false
+
+  # Override the vendor part of the Rust target triple (with a leading dash)
+  # used for building target code (not host code). Leave empty to use the
+  # platform default.
+  rust_target_triple_vendor_for_target = ""
 }
 
 declare_args() {
@@ -187,14 +192,21 @@ if (enable_rust) {
 # https://issues.chromium.org/u/1/issues/372512092#comment5 for an example.
 rust_abi_target = ""
 if (is_linux || is_chromeos) {
+  vendor = "-unknown"
+
+  is_host = current_toolchain == host_toolchain || current_toolchain == host_system_allocator_toolchain || toolchain_for_rust_host_build_tools
+  if (!is_host && rust_target_triple_vendor_for_target != "") {
+    vendor = rust_target_triple_vendor_for_target
+  }
+
   if (current_cpu == "arm64") {
-    rust_abi_target = "aarch64-unknown-linux-gnu"
+    rust_abi_target = "aarch64" + vendor + "-linux-gnu"
     cargo_target_abi = ""
   } else if (current_cpu == "x86") {
-    rust_abi_target = "i686-unknown-linux-gnu"
+    rust_abi_target = "i686" + vendor + "-linux-gnu"
     cargo_target_abi = ""
   } else if (current_cpu == "x64") {
-    rust_abi_target = "x86_64-unknown-linux-gnu"
+    rust_abi_target = "x86_64" + vendor + "-linux-gnu"
     cargo_target_abi = ""
   } else if (current_cpu == "arm") {
     if (arm_float_abi == "hard") {
@@ -213,18 +225,14 @@ if (is_linux || is_chromeos) {
       # The thumbv7 vs. armv7 distinction is for legacy reasons and both
       # targets in fact target Thumb, see:
       # https://github.com/rust-lang/rust/issues/44722
-      if (arm_use_neon) {
-        rust_abi_target = "thumbv7neon-unknown-linux-gnueabi" + float_suffix
-      } else {
-        rust_abi_target = "armv7-unknown-linux-gnueabi" + float_suffix
-      }
+      rust_abi_target = "armv7" + vendor + "-linux-gnueabi" + float_suffix
       cargo_target_abi = "eabi" + float_suffix
     } else {
       rust_abi_target = "arm-unknown-linux-gnueabi" + float_suffix
       cargo_target_abi = "eabi" + float_suffix
     }
   } else if (current_cpu == "riscv64") {
-    rust_abi_target = "riscv64gc-unknown-linux-gnu"
+    rust_abi_target = "riscv64gc" + vendor + "-linux-gnu"
     cargo_target_abi = ""
   } else if (current_cpu == "ppc64") {
     rust_abi_target = "powerpc64le-unknown-linux-gnu"
@@ -237,7 +245,7 @@ if (is_linux || is_chromeos) {
     cargo_target_abi = ""
   } else {
     # Best guess for other future platforms.
-    rust_abi_target = current_cpu + "-unknown-linux-gnu"
+    rust_abi_target = current_cpu + vendor + "-linux-gnu"
     cargo_target_abi = ""
   }
 } else if (is_android) {
diff --git a/build/rust/filter_clang_args.py b/build/rust/filter_clang_args.py
index 5a1843c0df..7fb4b49516 100644
--- a/build/rust/filter_clang_args.py
+++ b/build/rust/filter_clang_args.py
@@ -24,6 +24,11 @@ def filter_clang_args(clangargs):
           pass
         elif args[i].startswith('-plugin-arg'):
           i += 2
+      # Filter out ARM-specific flags that aren't supported on other architectures
+      elif args[i].startswith('-mbranch-protection'):
+        pass  # Skip this argument
+      elif args[i].startswith('-mfpu='):
+        pass
       else:
         yield args[i]
       i += 1
diff --git a/build/rust/rustc_wrapper.py b/build/rust/rustc_wrapper.py
index 8f2096dfe5..de43d44eed 100755
--- a/build/rust/rustc_wrapper.py
+++ b/build/rust/rustc_wrapper.py
@@ -160,6 +160,7 @@ def main():
   rustc_args = remaining_args[:ldflags_separator]
   ldflags = remaining_args[ldflags_separator + 1:rustenv_separator]
   rustenv = remaining_args[rustenv_separator + 1:sources_separator]
+  rustenv += ["RUSTC_BOOTSTRAP=1"]
 
   abs_build_root = os.getcwd().replace('\\', '/') + '/'
   is_windows = sys.platform == 'win32' or args.target_windows
diff --git a/build/rust/std/BUILD.gn b/build/rust/std/BUILD.gn
index bb2c988452..fdddf68d46 100644
--- a/build/rust/std/BUILD.gn
+++ b/build/rust/std/BUILD.gn
@@ -168,7 +168,8 @@ if (toolchain_has_rust) {
   # our locally-built std. Both reside in root_out_dir: we must only have one of
   # each per GN toolchain anyway.
 
-  sysroot_lib_subdir = "lib/rustlib/$rust_abi_target/lib"
+  sysroot_target_subdir = "lib/rustlib/$rust_abi_target"
+  sysroot_lib_subdir = "$sysroot_target_subdir/lib"
 
   if (!rust_prebuilt_stdlib) {
     local_rustc_sysroot = "$root_out_dir/local_rustc_sysroot"
@@ -295,12 +296,12 @@ if (toolchain_has_rust) {
         rust_abi_target,
       ]
 
-      outputs = []
+      outputs = [ "$target_out_dir/target.json" ]
       foreach(lib, all_stdlibs_to_copy) {
-        outputs += [ "$target_out_dir/lib$lib.rlib" ]
+        outputs += [ "$target_out_dir/lib/lib$lib.rlib" ]
       }
       foreach(lib, extra_sysroot_libs) {
-        outputs += [ "$target_out_dir/$lib" ]
+        outputs += [ "$target_out_dir/lib/$lib" ]
       }
 
       visibility = [ ":*" ]
@@ -313,8 +314,18 @@ if (toolchain_has_rust) {
                  "enable_rust=false")
       deps = [ ":find_stdlib" ]
       sources = get_target_outputs(":find_stdlib")
-      outputs =
-          [ "$prebuilt_rustc_sysroot/$sysroot_lib_subdir/{{source_file_part}}" ]
+      sources -= [ "$target_out_dir/target.json" ]
+      outputs = [
+        "$prebuilt_rustc_sysroot/$sysroot_lib_subdir/{{source_file_part}}",
+      ]
+
+      visibility = [ ":*" ]
+    }
+
+    copy("prebuilt_rustc_copy_target_json_to_sysroot") {
+      deps = [ ":find_stdlib" ]
+      sources = [ "$target_out_dir/target.json" ]
+      outputs = [ "$prebuilt_rustc_sysroot/$sysroot_target_subdir/target.json" ]
 
       visibility = [ ":*" ]
     }
@@ -354,7 +365,10 @@ if (toolchain_has_rust) {
         ":prebuilt_stdlib_libs",
         ":stdlib_public_dependent_libs",
       ]
-      deps = [ ":prebuilt_rustc_copy_to_sysroot" ]
+      deps = [
+        ":prebuilt_rustc_copy_to_sysroot",
+        ":prebuilt_rustc_copy_target_json_to_sysroot",
+      ]
     }
   }
 }
diff --git a/build/rust/std/find_std_rlibs.py b/build/rust/std/find_std_rlibs.py
index 386258f890..e8fdaa904a 100755
--- a/build/rust/std/find_std_rlibs.py
+++ b/build/rust/std/find_std_rlibs.py
@@ -17,7 +17,7 @@ import re
 from collections import defaultdict
 
 EXPECTED_STDLIB_INPUT_REGEX = re.compile(r"([0-9a-z_]+)(?:-([0-9]+))?$")
-RLIB_NAME_REGEX = re.compile(r"lib([0-9a-z_]+)-([0-9a-f]+)\.rlib$")
+RLIB_NAME_REGEX = re.compile(r"lib([0-9a-z_]+)(-([0-9a-f]+))?\.rlib$")
 
 
 def main():
@@ -52,6 +52,8 @@ def main():
     rustc_args.extend(["--target", args.target])
   rustlib_dir = subprocess.check_output(rustc_args).rstrip().decode()
 
+  lib_output_dir = os.path.join(args.output, 'lib')
+
   # Copy the rlibs to a predictable location. Whilst we're doing so,
   # also write a .d file so that ninja knows it doesn't need to do this
   # again unless the source rlibs change.
@@ -63,7 +65,7 @@ def main():
     # output rlibs for that purpose. If any of the input rlibs change, ninja
     # will run this script again and we'll copy them all afresh.
     depfile.write(
-        "%s:" % (os.path.join(args.output, "lib%s.rlib" % args.depfile_target)))
+        "%s:" % (os.path.join(lib_output_dir, "lib%s.rlib" % args.depfile_target)))
 
     def copy_file(infile, outfile):
       depfile.write(f" {infile}")
@@ -99,7 +101,7 @@ def main():
       # the correct file path to our linker invocations, we don't need
       # that, and it would prevent us having the predictable filenames
       # which we need for statically computable gn dependency rules.
-      (crate_name, metadata) = RLIB_NAME_REGEX.match(f).group(1, 2)
+      (crate_name, metadata) = RLIB_NAME_REGEX.match(f).group(1, 3)
 
       # Use the number of times we've seen this name to disambiguate the output
       # filenames. Since we sort the input filenames including the metadata,
@@ -117,14 +119,18 @@ def main():
       output_filename = f"lib{concise_name}.rlib"
 
       infile = os.path.join(rustlib_dir, f)
-      outfile = os.path.join(args.output, output_filename)
+      outfile = os.path.join(lib_output_dir, output_filename)
       copy_file(infile, outfile)
 
     for f in extra_libs:
       infile = os.path.join(rustlib_dir, f)
-      outfile = os.path.join(args.output, f)
+      outfile = os.path.join(lib_output_dir, f)
       copy_file(infile, outfile)
 
+    infile = os.path.join(os.environ['RUST_TARGET_PATH'], f'{args.target}.json')
+    outfile = os.path.join(args.output, 'target.json')
+    copy_file(infile, outfile)
+
     depfile.write("\n")
 
 
-- 
2.39.5

